### DISCUSSION TITLE
Improve error message

### DISCUSSION THREAD

**Discussion by Subhajit97** [microsoftgraph/msgraph-sdk-go#218] [Q&A] [answered]
Hello!
Is there any way to get a proper error message and code if API received any error?
I was trying to get a user by ID and passed an invalid ID.

Instead of giving a `ResourceNotFound` or `404` or similar error, it is returning:
```
Error ----> error status code received from the API
```

Code:
```
cred, err := azidentity.NewAzureCLICredential(
	&azidentity.AzureCLICredentialOptions{},
)

auth, err := a.NewAzureIdentityAuthenticationProvider(cred)
if err != nil {
	panic(err)
}

adapter, err := msgraphsdkgo.NewGraphRequestAdapter(auth)
if err != nil {
	panic(err)
}
client := msgraphsdkgo.NewGraphServiceClient(adapter)

input := &item.UserItemRequestBuilderGetQueryParameters{}
input.Select = []string{"displayName,Id"}
options := &item.UserItemRequestBuilderGetRequestConfiguration{
	QueryParameters: input,
}

user, err := client.UsersById("7e82c222-21e9-4339-bd36-83326cb1059g").GetWithRequestConfigurationAndResponseHandler(options, nil)
if err != nil {
	fmt.Printf("Error ----> %v", err)
	return
}

fmt.Println(*user.GetDisplayName(), *user.GetId())
```

I am using the SDK `v0.30.0`.

**Comment by baywet** [comment, marked as answer]
Hi @Subhajit97 
Thanks for using the Go SDK and for reaching out.
We've just had a community contribution yesterday that documents how to getter better information from the errors are you can see [here](https://github.com/microsoftgraph/msgraph-sdk-go#31-get-the-users-drive)

```Golang
import (
    "github.com/microsoftgraph/msgraph-sdk-go/models/odataerrors"
)

result, err := client.Me().Drive().Get()
if err != nil {
    fmt.Printf("Error getting the drive: %v\n", err)
    printOdataError(err)
}
fmt.Printf("Found Drive : %v\n", *result.GetId())

// omitted for brevity

func printOdataError(err error) {
	switch err.(type) {
	case *odataerrors.ODataError:
		typed := err.(*odataerrors.ODataError)
		fmt.Printf("error:", typed.Error())
		if terr := typed.GetError(); terr != nil {
			fmt.Printf("code: %s", *terr.GetCode())
			fmt.Printf("msg: %s", *terr.GetMessage())
		}
	default:
		fmt.Printf("%T > error: %#v", err, err)
	}
}
```

**Reply by Subhajit97** [reply]
Great! Thanks, @baywet, for the quick help.

**Comment by stephenwan-opal** [comment]
hi @baywet, i had a followup question: is matching against the string message the best way to check for specific error types? i.e.
```golang
if err := typed.GetError(); err != nil &&
  *err.GetCode() == "Request_BadRequest" &&
  strings.Contains(*err.GetMessage(), "One or more added object references already exist for the following modified properties") {
  // ... do something
}
```

I didn't see a more stable looking error code available in the returned type..

thanks!

**Reply by baywet** [reply]
Hi @stephenwan-opal
Thanks for the question.
Unfortunately this is a limitation of the service design. The error codes are a continuous collection (any API can introduce new codes even if it's discouraged).
So comparing string values of the code (and code of the inner error) is probably your best bet. I wouldn't compare the message however and I would use case insensitive comparison.
You can find more information about the most common error codes [here](https://learn.microsoft.com/en-us/graph/errors)

**Reply by stephenwan-opal** [reply]
oh, interesting - you don't recommend matching on the message at all?

My use case is that I'm trying to catch an error that throws when we try to do something that's fine for idempotency (adding a user to a group that they are already in), so I want to catch this specific error (not just any `Request_BadRequest`).

When I dump the error object out with `spew.Sdump`, I get:
```go
(*odataerrors.ODataError)(0x140005b8240)(error status code received from the API)
 map[] (abstractions.ApiError) error status code received from the API
 (*odataerrors.MainError)(0x1400014e3c0)({
 additionalData: (map[string]interface {}) (len=1) {
  (string) (len=10) "innerError": (map[string]*jsonserialization.JsonParseNode) (len=3) {
   (string) (len=4) "date": (*jsonserialization.JsonParseNode)(0x14000112250)({
    value: (*string)(0x14000112240)((len=19) "...")
   }),
   (string) (len=10) "request-id": (*jsonserialization.JsonParseNode)(0x140001122b0)({
    value: (*string)(0x140001122a0)((len=36) "...")
   }),
   (string) (len=17) "client-request-id": (*jsonserialization.JsonParseNode)(0x14000112310)({
    value: (*string)(0x14000112300)((len=36) "...")
   })
  }
 },
 code: (*string)(0x14000112160)((len=18) "Request_BadRequest"),
 details: ([]odataerrors.ErrorDetailsable) <nil>,
 innererror: (odataerrors.InnerErrorable) <nil>,
 message: (*string)(0x140001121c0)((len=99) "One or more added object references already exist for the following modified properties: 'members'."),
 target: (*string)(<nil>)
})
```

I don't see any other information on the error to match against except for the message (note that `innererror`, etc is nil) - is there something else to compare to get this specific error?

**Reply by baywet** [reply]
Yes some services don't always return an inner error. I'd suggest heading over to aka.ms/graphfeedback to advocate for more actionable errors.

**Reply by stephenwan-opal** [reply]
got it - thanks for the help and quick replies!

**Reply by baywet** [reply]
no worries, having a second look at the dump you shared it seems there's a deserialization issue with the inner error (you can see an entry in the additional data). Would you mind creating an issue please so we can look into that?

**Reply by stephenwan-opal** [reply]
filed here: https://github.com/microsoftgraph/msgraph-sdk-go/issues/334

thanks!
